Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -0,0 +1,215 @@
+    StrictMono (matricesOver (R := R) n) :=
+  fun I J IJ => by
+    refine lt_of_le_of_ne (matricesOver_monotone n IJ.le) fun eq => ?_
+    have ⟨x, xJ, xI⟩ : ∃ x ∈ J, x ∉ I := Set.exists_of_ssubset IJ
+    simp only [SetLike.mem_coe] at xI xJ
+    let M : Matrix n n R := fun _ _ => x
+    have : M ∈ J.matricesOver n := by simp [M, xJ]
+    have : M ∉ I.matricesOver n := by
+      simp only [mem_matricesOver, not_forall]
+      use h.some, h.some
+    simp [*] at *
+
+@[simp]
+theorem matricesOver_bot : (⊥ : Ideal R).matricesOver n = ⊥ := by
+  ext M
+  simp only [mem_matricesOver, mem_bot]
+  constructor
+  · intro H; ext; apply H
+  · intro H; simp [H]
+
+@[simp]
+theorem matricesOver_top : (⊤ : Ideal R).matricesOver n = ⊤ := by
+  ext; simp
+
+end Ideal
+
+/-! ### Two-sided ideals in a matrix ring -/
+
+namespace TwoSidedIdeal
+open Matrix
+
+variable {R : Type*} [Ring R]
+         (n : Type*) [Fintype n]
+
+/-- The two-sided ideal of matrices with entries in `I ≤ R`. -/
+def matricesOver (I : TwoSidedIdeal R) : TwoSidedIdeal (Matrix n n R) :=
+  TwoSidedIdeal.mk' { M | ∀ i j, M i j ∈ I }
+    (fun _ _ => I.zero_mem)
+    (fun ha hb i j => I.add_mem (ha i j) (hb i j))
+    (fun ha i j => I.neg_mem (ha i j))
+    (fun ha i j => by
+      rw [mul_apply]
+      apply sum_mem
+      intro k _
+      apply I.mul_mem_left _ _ (ha k j))
+    (fun ha i j => by
+      rw [mul_apply]
+      apply sum_mem
+      intro k _
+      apply I.mul_mem_right _ _ (ha i k))
+
+@[simp]
+lemma mem_matricesOver (I : TwoSidedIdeal R) (M : Matrix n n R) :
+    M ∈ I.matricesOver n ↔ ∀ i j, M i j ∈ I := by
+  simp [matricesOver]
+
+theorem matricesOver_monotone : Monotone (matricesOver (R := R) n) :=
+  fun _ _ IJ _ MI i j => IJ (MI i j)
+
+theorem matricesOver_strictMono_of_nonempty [h : Nonempty n] :
+    StrictMono (matricesOver (R := R) n) :=
+  fun I J IJ => by
+    refine lt_of_le_of_ne (matricesOver_monotone n IJ.le) fun eq => ?_
+    have ⟨x, xJ, xI⟩ : ∃ x ∈ J, x ∉ I := Set.exists_of_ssubset IJ
+    simp only [SetLike.mem_coe] at xI xJ
+    let M : Matrix n n R := fun _ _ => x
+    have : M ∈ J.matricesOver n := by simp [M, xJ]
+    have : M ∉ I.matricesOver n := by
+      simp only [mem_matricesOver, not_forall]
+      use h.some, h.some
+    simp [*] at *
+
+@[simp]
+theorem matricesOver_bot : (⊥ : TwoSidedIdeal R).matricesOver n = ⊥ := by
+  ext M
+  simp only [mem_matricesOver, mem_bot]
+  constructor
+  · intro H; ext; apply H
+  · intro H; simp [H]
+
+@[simp]
+theorem matricesOver_top : (⊤ : TwoSidedIdeal R).matricesOver n = ⊤ := by
+  ext; simp
+
+theorem asIdeal_matricesOver [DecidableEq n] (I : TwoSidedIdeal R) :
+    asIdeal (I.matricesOver n) = (asIdeal I).matricesOver n := by
+  ext; simp
+
+variable {n : Type*} [Fintype n] [DecidableEq n]
+
+/--
+Two-sided ideals in `R` correspond bijectively to those in `Mₙ(R)`.
+Given an ideal `I ≤ R`, we send it to `Mₙ(I)`.
+Given an ideal `J ≤ Mₙ(R)`, we send it to `{Nᵢⱼ | ∃ N ∈ J}`.
```

Comment:
As a general comment, I'd recommend using `$Mₙ(R)$` for anything that isn't actually valid lean notation, since that will render more nicely in the docs, and avoid any confusion about why code in docstrings doesn't actually work.